src/server.rs

```diff
@@ -78,12 +78,34 @@ impl conmon::Server for ConmonServerImpl {
     }
 }
 
-// Use the single threaded runtime to save rss memory
-#[tokio::main(flavor = "current_thread")]
-async fn main() -> Result<(), Error> {
-    let (shutdown_tx, shutdown_rx) = oneshot::channel();
-    let server = ConmonServerImpl::new().await?;
+fn main() -> Result<(), Error> {
+    // First, initialize the server so we have access to the config pre-fork.
+    let server = ConmonServerImpl::new()?;
+
+    // We need to fork as early as possible, especially before setting up tokio.
+    // If we don't, the child will have a strange thread space and we're at risk of deadlocking.
+    // We also have to treat the parent as the child (as described in [1]) to ensure we don't
+    // interrupt the child's execution.
+    // 1: https://docs.rs/nix/0.23.0/nix/unistd/fn.fork.html#safety
+    match unsafe { fork() } {
+        Ok(ForkResult::Parent { child, .. }) => {
+            if let Some(path) = server.config().conmon_pidfile() {
+                let child_str = format!("{}", child);
+                File::create(path)?.write_all(child_str.as_bytes())?;
+            }
+            unsafe { _exit(0) };
```

Comment:
I can't figure out if it's [async signal safe](https://docs.rs/nix/0.23.0/nix/unistd/fn.fork.html#safety)

Reply:
It's hard to know for sure; I don't think Rust makes strong promises about what system calls the standard library could make.  For example, `open()/write()/close()` corresponding to the `File::create().write_all()` calls above are safe...but in theory `std` could switch to some other system call.  Though that seems very unlikely.

As discussed on an internal chat I think really the best bet is to avoid a bare `fork()` at all and instead e.g. have podman use `systemd-run` or a direct call to the DBus `StartTransientUnit` API.  

Or to be really sure, only invoke `libc` API in the forked child - in the same way one would do in C.

But anyways, I think this code is all fine as is really too.  I'd keep the `_exit()` over `std::process::exit()`.

Reply:
FWIW, the subprocess crate uses [fork and then raw exit](https://github.com/hniksic/rust-subprocess/blob/f092a654d6668736f0ce0f52fb31e17f3e31e091/src/popen.rs#L744).

Reply:
After more thought and discussion with @mrunalp , I think it'd be best to go forward with the daemonize code for now. it's been proven to work with conmon, and it will allow the port to happen in a cleaner way. in the future, we can revisit that decision